{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2023-04-24\n",
      "Gurobi Optimizer version 9.5.2 build v9.5.2rc0 (mac64[x86])\n",
      "Thread count: 4 physical cores, 8 logical processors, using up to 8 threads\n",
      "Optimize a model with 93 rows, 45 columns and 168 nonzeros\n",
      "Model fingerprint: 0xfb8c732f\n",
      "Coefficient statistics:\n",
      "  Matrix range     [1e+00, 1e+00]\n",
      "  Objective range  [1e+00, 1e+00]\n",
      "  Bounds range     [0e+00, 0e+00]\n",
      "  RHS range        [4e+00, 4e+00]\n",
      "Presolve removed 40 rows and 8 columns\n",
      "Presolve time: 0.00s\n",
      "\n",
      "Solved in 0 iterations and 0.00 seconds (0.00 work units)\n",
      "Infeasible model\n",
      "C0\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "Unable to retrieve attribute 'X'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m/Volumes/Data/Dokumente/Uni/Bioinformatik/2.Semester/FortgeschritteneMethodenderBioinformatik/Projekt/GIT-Hub/Transkript-Assembler/Src/main.ipynb Zelle 1\u001b[0m in \u001b[0;36m<cell line: 20>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Volumes/Data/Dokumente/Uni/Bioinformatik/2.Semester/FortgeschritteneMethodenderBioinformatik/Projekt/GIT-Hub/Transkript-Assembler/Src/main.ipynb#W0sZmlsZQ%3D%3D?line=41'>42</a>\u001b[0m     G_clean \u001b[39m=\u001b[39m G_full\n\u001b[1;32m     <a href='vscode-notebook-cell:/Volumes/Data/Dokumente/Uni/Bioinformatik/2.Semester/FortgeschritteneMethodenderBioinformatik/Projekt/GIT-Hub/Transkript-Assembler/Src/main.ipynb#W0sZmlsZQ%3D%3D?line=43'>44</a>\u001b[0m transcripts \u001b[39m=\u001b[39m []\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Volumes/Data/Dokumente/Uni/Bioinformatik/2.Semester/FortgeschritteneMethodenderBioinformatik/Projekt/GIT-Hub/Transkript-Assembler/Src/main.ipynb#W0sZmlsZQ%3D%3D?line=44'>45</a>\u001b[0m g_Star \u001b[39m=\u001b[39m writeGStar(G_full)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Volumes/Data/Dokumente/Uni/Bioinformatik/2.Semester/FortgeschritteneMethodenderBioinformatik/Projekt/GIT-Hub/Transkript-Assembler/Src/main.ipynb#W0sZmlsZQ%3D%3D?line=45'>46</a>\u001b[0m plt\u001b[39m.\u001b[39mfigure(\u001b[39m1\u001b[39m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Volumes/Data/Dokumente/Uni/Bioinformatik/2.Semester/FortgeschritteneMethodenderBioinformatik/Projekt/GIT-Hub/Transkript-Assembler/Src/main.ipynb#W0sZmlsZQ%3D%3D?line=46'>47</a>\u001b[0m nx\u001b[39m.\u001b[39mdraw(g_Star, with_labels\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m/Volumes/Data/Dokumente/Uni/Bioinformatik/2.Semester/FortgeschritteneMethodenderBioinformatik/Projekt/GIT-Hub/Transkript-Assembler/Src/flowProblem.py:94\u001b[0m, in \u001b[0;36mwriteGStar\u001b[0;34m(graph)\u001b[0m\n\u001b[1;32m     92\u001b[0m \u001b[39mfor\u001b[39;00m var \u001b[39min\u001b[39;00m flowModel\u001b[39m.\u001b[39mgetVars():\n\u001b[1;32m     93\u001b[0m     \u001b[39mprint\u001b[39m(var\u001b[39m.\u001b[39mvarName)\n\u001b[0;32m---> 94\u001b[0m     \u001b[39mprint\u001b[39m(var\u001b[39m.\u001b[39;49mX)\n\u001b[1;32m     95\u001b[0m \u001b[39mprint\u001b[39m(flow)\n\u001b[1;32m     96\u001b[0m \u001b[39mreturn\u001b[39;00m(graphStar)\n",
      "File \u001b[0;32msrc/gurobipy/var.pxi:125\u001b[0m, in \u001b[0;36mgurobipy.Var.__getattr__\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32msrc/gurobipy/var.pxi:153\u001b[0m, in \u001b[0;36mgurobipy.Var.getAttr\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32msrc/gurobipy/attrutil.pxi:100\u001b[0m, in \u001b[0;36mgurobipy.__getattr\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: Unable to retrieve attribute 'X'"
     ]
    }
   ],
   "source": [
    "# IMPORT\n",
    "import sys\n",
    "import parse_graph_new\n",
    "import path_enumeration\n",
    "import pairedbin_enumeration\n",
    "import networkx as nx\n",
    "import time\n",
    "import matplotlib.pyplot as plt\n",
    "from copy import deepcopy \n",
    "import optimize\n",
    "from flowProblem import writeGStar\n",
    "\n",
    "# VARIABLES\n",
    "start = time.time()\n",
    "no_trans = 0\n",
    "#gene_id = 0\n",
    "#file_gtf = open(\"transcripts.gtf\", \"w\") \n",
    "\n",
    "# MAIN\n",
    "with open('test1.graph') as f:\n",
    "    fileEndReached = False\n",
    "    f.readline() #skip ---- seperator line\n",
    "    while not fileEndReached:\n",
    "        \n",
    "        # READ META AND BIN DATA FROM FILE\n",
    "        f.readline() #skip ==META \n",
    "        Chromosome, Strand, Exons = parse_graph_new.parse_meta(f)\n",
    "        Bins = parse_graph_new.parse_bins(f)\n",
    "        PairedBins = parse_graph_new.parse_pairs(f)\n",
    "        PairedBins_copy = deepcopy(PairedBins)\n",
    "\n",
    "        # BUILD GRAPH\n",
    "        G_full = nx.DiGraph()\n",
    "        fileEndReached, skip = parse_graph_new.parse_graph(f, G_full, Exons)\n",
    "\n",
    "        if not fileEndReached and not skip:\n",
    "            G_clean = nx.DiGraph()\n",
    "            fileEndReached, _ = parse_graph_new.parse_graph(f, G_clean, Exons)\n",
    "            # nx.draw_networkx(G_clean, with_labels=True, arrowsize=12)\n",
    "            # plt.show()\n",
    "        if skip:\n",
    "            G_clean = G_full\n",
    "\n",
    "        transcripts = []\n",
    "        g_Star = writeGStar(G_full)\n",
    "        plt.figure(1)\n",
    "        nx.draw(g_Star, with_labels=True)\n",
    "        plt.figure(2)\n",
    "        nx.draw(G_full, with_labels=True)\n",
    "        # FULL PATH ENUMERATION\n",
    "        \"\"\"\n",
    "        transcripts = path_enumeration.enumeration(G_full,[],\"0\",[\"0\"],\"1\",True)\n",
    "        no_trans = no_trans + len(transcripts)\n",
    "        \"\"\"\n",
    "        \n",
    "        # MULTI BIN ENUMERATION\n",
    "        \"\"\"\n",
    "        multi_bins = path_enumeration.get_multibins(Bins)\n",
    "        transcripts = path_enumeration.enumeration_bins2(G_clean,[],\"0\",[\"0\"],[],multi_bins,\"1\",True)\n",
    "        no_trans = no_trans + len(transcripts)\n",
    "        \"\"\"\n",
    "        \n",
    "        # PAIRED BIN ENUMERATION 1\n",
    "        \"\"\"\n",
    "        multi_bins = path_enumeration.get_multibins(Bins)\n",
    "        paired_bins = pairedbin_enumeration.get_pairedbins(G_clean,PairedBins_copy,multi_bins)\n",
    "        transcripts = path_enumeration.enumeration_bins2(G_clean,[],\"0\",[\"0\"],[],paired_bins+multi_bins,\"1\",True)\n",
    "        no_trans = no_trans + len(transcripts)\n",
    "        \"\"\"\n",
    "        \n",
    "        # PAIRED BIN ENUMERATION 2\n",
    "        \"\"\"\n",
    "        pairedbins_grouped = pairedbin_enumeration.group_pairs(PairedBins_copy)\n",
    "        multi_bins = path_enumeration.get_multibins(Bins)\n",
    "        transcripts = path_enumeration.enumeration_bins2(G_clean,[],\"0\",[\"0\"],[],multi_bins,\"1\",True)\n",
    "        transcripts_copy = deepcopy(transcripts)\n",
    "        filtered_transcripts = pairedbin_enumeration.filter_transcripts(transcripts_copy,pairedbins_grouped)\n",
    "        no_trans = no_trans + len(filtered_transcripts)\n",
    "        \"\"\"\n",
    "        \n",
    "        # ADD TRANSCRIPTS TO GTF FILE\n",
    "        \"\"\"\n",
    "        gene_id += 1\n",
    "        transcript_id = 0\n",
    "        for transcript in transcripts:\n",
    "            transcript_id += 1\n",
    "            parse_graph_new.write_valid_gtf_entry(file_gtf,Chromosome,Strand,Exons,transcript,\"Gene\"+str(gene_id),\"Transcript\"+str(transcript_id))\n",
    "        \"\"\"\n",
    "    \n",
    "        # Optimization \n",
    "        \"\"\"\n",
    "        transcripts = path_enumeration.enumeration(G_clean, [], \"0\", [\"0\"], \"1\", False)\n",
    "        print(optimize.model(G_clean, transcripts))\n",
    "        #get dictionary with expression levels (F=(f1,f2,f3,...fN))\n",
    "        var_dict = optimize.model(G_clean, transcripts)\n",
    "        \"\"\"\n",
    "\n",
    "\n",
    "    # PRINT RESULTS  \n",
    "    end = time.time()\n",
    "    print(\"Gesamtanzahl Transkripte: \", no_trans)\n",
    "    print('{:5.3f}s'.format(end-start))       \n",
    "    #file_gtf.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.5 ('TranscriptReconstruction')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "8dd36f0af0ffccb3fea10f75b284f29e0157116db7216fa38748737214da1580"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
