{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GroupPairedBinT(leftExons=[0, 2], rightExons=[[2, 5]], count={'1': 1, '3': 2, 'c': 3})\n",
      "GroupPairedBinT(leftExons=[0], rightExons=[[2]], count=[{'4': 1, 'c': 1}])\n",
      "GroupPairedBinT(leftExons=[1, 2], rightExons=[[2, 3], [2, 5], [3], [4, 5], [5]], count=[{'2': 1, '0': 2, '4': 3, '1': 2, '3': 12, 'c': 20}, {'1': 3, '4': 8, '3': 8, '2': 12, '0': 11, 'c': 42}, {'4': 1, '3': 1, 'c': 2}, {'1': 1, 'c': 1}, {'0': 1, '3': 1, '1': 1, '2': 2, 'c': 5}])\n",
      "GroupPairedBinT(leftExons=[1], rightExons=[[2, 3], [2, 5]], count=[{'0': 1, 'c': 1}, {'3': 1, '4': 1, '2': 1, '0': 1, 'c': 4}])\n",
      "GroupPairedBinT(leftExons=[2, 5], rightExons=[[5, 6]], count=[{'3': 1, '2': 3, '4': 1, 'c': 5}])\n",
      "GroupPairedBinT(leftExons=[2], rightExons=[[5]], count=[{'3': 7, '1': 6, '2': 8, '0': 9, '4': 7, 'c': 37}])\n",
      "GroupPairedBinT(leftExons=[5, 7], rightExons=[[7, 8]], count=[{'0': 2, 'c': 2}])\n",
      "GroupPairedBinT(leftExons=[5], rightExons=[[7, 8]], count=[{'3': 3, '1': 2, '4': 2, '2': 7, '0': 11, 'c': 25}])\n"
     ]
    }
   ],
   "source": [
    "import sys, ast, os\n",
    "import networkx as nx\n",
    "from collections import namedtuple\n",
    "from PathEnumeration import fullPathEnumeration, activeBinPathEnumeration, activeBinPathEnumeration2, activeBinPathEnumeration3, getMultiBins\n",
    "from PathEnumeration_old import activeMultiBinPathEnumeration\n",
    "from PairedBinsToBins import fromPairedBinsToBins, checkPairedBins\n",
    "from PostFilterPairedBin import groupPairedBins, eliminateInvalidPaths\n",
    "from PairedBinsToBins_short import fromPairedBinsToBinsShort\n",
    "from parse_graph_list_commented_Arbeitsdatei import parse_meta, parse_bins, parse_pairs, parse_graph, write_valid_gtf_entry, nodepath_to_transcript\n",
    "from copy import deepcopy\n",
    "\n",
    "dummyf = open(\"dummyout.gtf\", \"w\")                                                                                              # output for the dummy code\n",
    "dummyGeneCounter = 0\n",
    "\n",
    "with open('test.graph') as f:\n",
    "    fileEndReached = False\n",
    "    f.readline()                                                                                                    #skip ---- seperator line\n",
    "    \n",
    "    # Define necessary Dictionaries                                                                                                     \n",
    "    paths = {}\n",
    "    filteredPaths = {}\n",
    "    pairedBinDict = {}\n",
    "\n",
    "    #Define necessary lists\n",
    "    pfadListe = []\n",
    "\n",
    "    # Define necessary Integer Varibales\n",
    "    geneCounter = 0\n",
    "    enumerationPathCounter = 0\n",
    "    validPathCounter = 0\n",
    "    \n",
    "    while not fileEndReached:\n",
    "        f.readline()                                                                                                # skip ==META: Read this line, but don't do anything\n",
    "        Chromosome, Strand, Exons = parse_meta(f)                                                                   # Übergib f jetzt an def parse_meta, um Metadaten auszulesen und schreib diese in Chromosome, \n",
    "                                                                                                                    # Strand und Exons (Listen)\n",
    "        Bins = parse_bins(f)                                                                                        # Lies die BINS aus f mit Hilfe der parse_bin(f) Funktion aus und schreib sie in Bins\n",
    "        PairedBins = parse_pairs(f)\n",
    "        \n",
    "        G_full = nx.DiGraph()                                                                                       # Erzeug einen Diagraphen mit Hilfe von NetworkX\n",
    "        fileEndReached, skip = parse_graph(f, G_full, Exons)                                                        # Ruf die Funktion parse_graph auf, übergibt ihr das File (f, den erzeugten Graphen und zugehörige Liste \n",
    "                                                                                                                    # mit den Exons), schreib den vollen Graphen in G_full, weise Skip einen Boolean zu, der true ist, wenn \n",
    "                                                                                                                    # die Zeile mit - beginnt                                                                                                                                                                                                      \n",
    "        if not fileEndReached and not skip:                                                                         # Falls denoised Graph existiert, ruf wieder die Funktion Parse_Graph auf, übergib ihr das File \n",
    "                                                                                                                    # (f, Graph_clean und zugehörige Liste mit den Exons), schreib die # denoised Informationen aus dem \n",
    "                                                                                                                    # Graph-File in graph clean und übergib die letzte Zeile fileEndReached\n",
    "            G_clean = nx.DiGraph()                                                                                  # Erzeug einen neuen gerichteten Graphen \n",
    "            fileEndReached, _ = parse_graph(f, G_clean, Exons)\n",
    "        \n",
    "        else:\n",
    "            G_clean = G_full\n",
    "\n",
    "        # Define necessary variables and assign\n",
    "            # pathNumbers to enumerate paths  \n",
    "        \n",
    "        enumerationPathNumber = [0]\n",
    "        validPathNumber = [0]\n",
    "        pairedBinsCopy = deepcopy(PairedBins)\n",
    "\n",
    "        # 1. Get MultiBins only from Bins\n",
    "        MultiBins = getMultiBins(Bins)\n",
    "\n",
    "        # 3. Add PairedBins to MultiBins\n",
    "        #MultiBins, gepaarteBins = fromPairedBinsToBinsShort(PairedBins, MultiBins, G_clean, Exons, True)\n",
    "        \n",
    "        # 4. Enumerate Paths with ActiveBinPathEnumeration3\n",
    "        paths['Gene' + str(geneCounter)] = activeBinPathEnumeration3('1', '0', ['0'], {}, enumerationPathNumber, [], G_clean, MultiBins)\n",
    "\n",
    "        # 4a Enumerate paths with ActiveBinPathEnumeration3 (Bins)\n",
    "        #paths['Gene' + str(geneCounter)] = activeBinPathEnumeration3('1', '0', ['0'], {}, enumerationPathNumber, [], G_clean, MultiBins)\n",
    "\n",
    "        # 5. Group paired Bins to filter valid paths after Enumeration \n",
    "        groupedPairedBins = groupPairedBins(pairedBinsCopy)\n",
    "\n",
    "        # 6. Filter invalid Paths\n",
    "        #filteredPaths['Gene' + str(geneCounter)] = eliminateInvalidPaths(paths['Gene' + str(geneCounter)], groupedPairedBins, validPathNumber)\n",
    "        \n",
    "        # 7. Increase geneCounter\n",
    "        geneCounter = geneCounter + 1\n",
    "\n",
    "        # 8. Count paths\n",
    "        enumerationPathCounter = enumerationPathCounter + enumerationPathNumber[0]\n",
    "        validPathCounter = validPathCounter + validPathNumber[0]\n",
    "        \n",
    "        # All Paths Enumeration\n",
    "\n",
    "        # Note: source and drain are ALWAYS named \"0\" and \"1\" respectively\n",
    "\n",
    "        #if skip:\n",
    "            # handle the rare case that noise deletion removes the whole second graph\n",
    "        \n",
    "\n",
    "        # TODO WORK WITH THE GRAPH HERE\n",
    "        #Access Edge Types : G.edges[n1 , n2]['type'] == \"Exon\" || \"SpliceJunction\" || \"Helper\"\n",
    "        #Access Main Coverage Count of an Edge : G.edges[n1 , n2]['counts']['c']\n",
    "        #Access Exon length G.edges[n1 , n2]['length']\n",
    "        \n",
    "        #Source Node s is always G.nodes['0']\n",
    "        #Drain Node t is always G.nodes['1']\n",
    "        \n",
    "        # DUMMY Code extracts longest Path (by number of bases) and writes it to a GTF file\n",
    "        \n",
    "\n",
    "        lpath = nx.dag_longest_path(G_full, weight=\"length\")\n",
    "        transcript = nodepath_to_transcript(G_full, lpath)\n",
    "        write_valid_gtf_entry(dummyf, Chromosome, Strand, Exons, transcript, \"Gene\"+str(dummyGeneCounter), \"Transcript\"+str(dummyGeneCounter)+\".1\")\n",
    "        dummyGeneCounter = dummyGeneCounter + 1\n",
    "\n",
    "dummyf.close()\n",
    "for bin in groupedPairedBins:\n",
    "    print(bin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "laurasFileEndReached = False\n",
    "laurasFileName = 'vergleich_pairedbins2.txt'\n",
    "laurasTranscriptList = []\n",
    "with open (laurasFileName, 'r') as laurasFile:\n",
    "    line = laurasFile.readline()\n",
    "    while (laurasFileEndReached==False):\n",
    "        laurasTranscriptList.append(int(line))\n",
    "        line = laurasFile.readline()\n",
    "        if line == '':\n",
    "            laurasFileEndReached = True\n",
    "\n",
    "print (len(laurasTranscriptList))\n",
    "print (len(filteredPaths.values()))\n",
    "richisTranscriptList = list(filteredPaths.values())\n",
    "erroneousTranscripts = []\n",
    "summe=0\n",
    "\n",
    "for i in range(0,len(laurasTranscriptList)):\n",
    "    if len(richisTranscriptList[i])!=laurasTranscriptList[i]:\n",
    "        erroneousTranscripts.append(i)\n",
    "        summe = summe + (laurasTranscriptList[i]-len(richisTranscriptList[i]))\n",
    "print(erroneousTranscripts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(laurasTranscriptList[14])\n",
    "print(len(richisTranscriptList[14]))\n",
    "print(paths['Gene14'])\n",
    "print(len(paths['Gene14']))\n",
    "print(filteredPaths['Gene14'])\n",
    "print(len(filteredPaths['Gene14']))\n",
    "print(summe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "31848095eba7989a684b10cb208c018790ea1a5b15d11c8d6c10ed7fd9c1b69a"
  },
  "kernelspec": {
   "display_name": "Python 3.9.12 ('TranscriptReconstruction')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
